### PR TITLE
add a form to save an image url

### PR DESCRIPTION
**Pull request by kevinkir**
Closes Issue 2: Save Image Link

### CODE REVIEW

**Comment by kevinkir** [comment]
[Issue 2: Save Image Link](https://trello.com/c/KroSCnqj/8-issue-2-save-image-link)

File: test/models/image_test.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+class ImageTest < ActiveSupport::TestCase
+  test 'rejects a blank url' do
+    image = Image.new(url: '   ')
+    assert_not image.valid?
+    assert_equal "Url can't be blank", image.errors.full_messages.first
+  end
+
+  test 'rejects a url which has an invalid protocol' do
+    image = Image.new(url: 'foo://asdf')
+    assert_not image.valid?
```
[the remark this answers is not in the document]
Comment (kevinkir) [reply test/models/image_test.rb:12]: Cool! Thanks for the tip

File: app/views/images/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<h1>Add an image</h1>
+<%= form_with(model: @image, local: true) do |form| %>
+  <% if @image.errors.any? %>
+    <div id="error_explanation">
+      <div class="alert alert-danger">
+        The form contains <%= pluralize(@image.errors.count, 'error') %>
```
[the remark this answers is not in the document]
Comment (kevinkir) [reply app/views/images/new.html.erb]: I went ahead and put this as a flash message. I still have the bullet points to provide more detail. I figured it might be nice to add the flexibility to show multiple error messages. Would it be more conventional to show all the error messages as part of the flash?

File: app/views/images/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<h1>Add an image</h1>
+<%= form_with(model: @image, local: true) do |form| %>
```
[the remark this answers is not in the document]
Comment (kevinkir) [reply app/views/images/new.html.erb:2]: 👍 I'll check that out. Thanks.

File: app/views/images/new.html.erb (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmangel) [reply app/views/images/new.html.erb]: actually more conventional is to show the error messages inline in the form, which you'll see if you look into simple_form

File: test/controllers/images_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    image.stubs(:save).returns(false)
+    Image.stubs(:new).returns(image)
+
+    post images_url, params: {
+      image: { url: 'http://bogus-url' }
+    }
+    assert_response :unprocessable_entity
+  end
+
+  test 'should show an image' do
```
Comment (jmangel) [inline test/controllers/images_controller_test.rb:32]: This is a good time to start thinking about testing edge cases. We don't want users to see a generic rails error page if they try to view an issue that has been deleted, do you think we should add a test for what happens when they try to show a non-existent image id? (for testing purposes, you can use `get image_url(id: -1)`

File: test/controllers/images_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
+require 'test_helper'
+
+class ImagesControllerTest < ActionDispatch::IntegrationTest
```
Comment (jmangel) [inline test/controllers/images_controller_test.rb:3]: 👍

File: test/controllers/images_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+  test 'should return an error response on create failure' do
+    image = Image.new
+    image.stubs(:save).returns(false)
+    Image.stubs(:new).returns(image)
+
+    post images_url, params: {
+      image: { url: 'http://bogus-url' }
+    }
+    assert_response :unprocessable_entity
```
Comment (jmangel) [inline test/controllers/images_controller_test.rb:29]: 👍

File: app/validators/image_url_validator.rb [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    rescue URI::InvalidURIError
+      return 'is invalid'
+    end
+
+    return 'must be http or https' unless parsed_url.is_a?(URI::HTTP)
+
+    begin
+      response = Net::HTTP.get_response(parsed_url)
+    # Catch-all here, as this can raise a lot of different errors.
+    rescue # rubocop:disable Style/RescueStandardError
```
Comment (jmangel) [inline app/validators/image_url_validator.rb:26]: I believe `rescue` rescues `StandardError` by default, so we might want to listen to the rubocop rule here. And I don't think we want to rescue anything broader than StandardError anyway https://thoughtbot.com/blog/rescue-standarderror-not-exception
In general I think we should be thoughtful about disabling rubocop errors, there are certainly cases where we don't need to follow them, but they are there for a reason (at least in preparation for when you get into APM)

File: config/routes.rb [hunk 8]
```diff
@@ -1,5 +1,5 @@
 Rails.application.routes.draw do
   get 'welcome/index'
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  resources :images
```
[the remark this answers is not in the document]
Comment (jmangel) [reply config/routes.rb:3]: Do you think we should make this change? I think this comment may have been hidden last time you looked, github does that when there are a lot of comments

File: app/views/welcome/index.html.erb [hunk 9]
```diff
@@ -1,2 +1,5 @@
-<h1>Welcome#index</h1>
-<p>Find me in app/views/welcome/index.html.erb</p>
+<h1>
+  <%= image_tag('camera.svg', alt: 'camera icon', width: 50) %>
+  Kevin's Image Sharer
+</h1>
+<%= link_to 'add an image', new_image_path %>
```
Comment (jmangel) [inline app/views/welcome/index.html.erb:5]: should we add a simple `index` test in the controller test as well?

File: app/views/images/new.html.erb (hunk 2 above)
[the remark this answers is not in the document]
Comment (kevinkir) [reply app/views/images/new.html.erb]: Ok, cool. I'm going to take a look at that later.
